README
Visual Studio project
====================================
Compile tools and swath projects from Visual Studio 2017


SWATH (Smart Word Analysis for THai)
====================================

Thai script has no word delimiter. While it's trivial for human readers to
recognize word boundaries while reading, it requires some knowledge for the
machine to do the same when wrapping lines or moving cursor word-wise, etc.
Normally, applications need such feature to support Thai text processing.

Swath is a general-purpose utility to workaround the lack of such capability
in applications. It analyzes the given Thai text by consulting a Thai word
list for word boundaries, before outputting the same text with the predefined
word delimiters inserted.

It can read many kinds of input, including plain text and structured documents
like HTML, RTF, LaTeX and Lambda (Unicode version of LaTeX with Omega
typesetter kernel). [See -f option].

For the known documents, it inserts the common word delimiters used in the
corresponding formats, and pipes (|) for plain text. But the user can always
override this with a preferred delimiter. [See -b option.]

Swath can also be configured to use different algorithms for the analysis.
Currently, it supports two schemes: longest (greedy) matching and maximal
(least words) matching. [See -m option.]

EXAMPLES
========

- For LaTeX (to be used with babel-thai package):

    $ swath -f latex < mydoc.tex > mydoc.ttex
    $ latex mydoc.ttex

  Or if you composed your LaTeX source in UTF-8:

    $ swath -f latex -u u,t mydoc.tex > mydoc.ttex
    $ latex mydoc.ttex

  This is equivalent to filtering with iconv(1):

    $ iconv -f UTF-8 -t TIS-620 mydoc.tex | swath -f latex > mydoc.ttex
    $ latex mydoc.ttex

- For HTML (to provide web pages to web browsers that cannot wrap Thai lines
  properly, but support the <wbr> tag):

    $ swath -f html < mydoc.html > mydoc-wbr.html




ChangeLog
2020-09-09  madrazo
	Added VS2017 compilation


2020-08-13  Theppitak Karoonboonyanan  <theppitak@gmail.com>

	Fix 'make distcheck' failure.

	* Makefile.am:
	  - Add 'lib' to SUBDIRS.

2020-08-06  Theppitak Karoonboonyanan  <theppitak@gmail.com>

	Fix missing wcpcpy() & wcpncpy() in NetBSD

	* configure.ac, +lib/, +lib/Makefile.am, +lib/wcpcpy.c:
	  - Check for wcpcpy() and provide a replacement if missing.
	* src/Makefile.am:
	  - Add $(LTLIBOBJS) to swath_LIBADD.
	  - Use 'subdir-objects' automake option.
	* src/filterrtf.cpp:
	  - Add optional declaration of wcpcpy() if missing.
	  - [FilterRTF::GetNextToken()] Replace wcpncpy() call with
	    wcsncpy() and manual pointer advancement, as the pre-calculated
	    nCopy already determines the exact characters copied.

	Thanks scole_mail for the report via a personal mail.

2019-12-20  Theppitak Karoonboonyanan  <theppitak@gmail.com>

	Use GitHub issue tracker as bug report address.

	* configure.ac:
	  - Replace bug report e-mail address with GitHub issue tracker URL.

2019-05-08  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Make help message clearer.

	* src/wordseg.cpp (Usage()):
	  - Split "-v" out of "mule".
	  - Also mention "--verbose".
	* src/wordseg.cpp (Usage(), main()):
	  - Add "-h" option.
	  - Also mention "--help".

	As part of pull request #5
	https://github.com/tlwg/swath/pull/5
	Thanks @pepa65

2019-05-08  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Add documentation on build requirements.

	* INSTALL:
	  - Add build requirements.

	As part of pull request #5
	https://github.com/tlwg/swath/pull/5
	Thanks @pepa65

2019-05-08  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Replace "delimitor" by "delimiter".

	* ChangeLog, README, src/swath.1, src/wordseg.cpp:
	  - Fix all typos "delimitor" with "delimiter".

	As part of pull request #5
	https://github.com/tlwg/swath/pull/5
	Thanks @pepa65

2018-08-20  Theppitak Karoonboonyanan <theppitak@gmail.com>

	* NEWS:
	  === Version 0.6.1 ===

2018-08-18  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Add info to skip a step on released tarballs

	* INSTALL:
	  - Add info to skip ./autogen.sh step on released tarballs.

2018-08-18  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Provide our own version of INSTALL instruction.

	* INSTALL:
	  - Explain simple installation steps.

	Merging pull request #3 from @pepa65
	https://github.com/tlwg/swath/pull/3

2018-08-14  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Fix compiler warning on uncovered enum case.

	* src/filterrtf.cpp (FilterRTF::GetNextToken()):
	  - Add default: to switch(), to fix warning on uncovered enum.
	  - Remove an unnecessary break after goto.

2018-08-14  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Add comment for previous commit.

	* src/filterrtf.cpp (FilterRTF::chgCharState()):
	  - Add comment explaining unreachable line.

2018-08-14  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Drop unused CS_NONE ECharState in FilterRTF.

	The only place it's used is the return value of chgCharState()
	method, which has no chance to be reached, as the 'state' arg,
	which is from FilterRTF::psState, is never set to CS_NONE
	through out the code. So, all possible values have been covered
	in 'switch (state)' and compiler warning on uncovered CS_NONE
	is in fact meaningless.

	Besides, having a "none" value for a state in finite automaton
	is questionable. Is it a special dead state? No, it's never
	handled explicitly anyways.

	* src/filterrtf.h (FilterRTF::ECharState):
	  - Drop CS_NONE value.
	* src/filterrtf.cpp (FilterRTF::chgCharState()):
	  - Drop unreachable "return CS_NONE" statement.

2018-08-13  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Fix inverted logic in RTFToken::isEmpty().

	* src/filterrtf.cpp (RTFToken::isEmpty()):
	  - Fix inverted logic, which caused some token to be not flushed.
	* tests/petavatthu1-wseg.rtf:
	  - Update checking wordseg output.

2018-08-13  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Add RTF test case.

	* tests/Makefile.am, +tests/test-rtf.sh,
	  +tests/petavatthu1.rtf, +tests/petavatthu1-wseg.rtf:
	  - Add RTF test case

2018-08-13  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Drop unused method.

	* src/filterrtf.cpp (-RTFToken::set(char, ETokenType)):
	  - Drop unused method.

2018-08-13  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Declare array size member as size_t.

	* src/filterrtf.cpp (RTFToken):
	  - Declare RTFToken::valLen as size_t instead of int,
	    fixing "signedness" compiler warning when comparing size.

2018-08-13  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Add missing virtual d-tors.

	* conv/convkit.h (TextReader, TextWriter):
	  - Add missing virtual d-tors to the abstract base classes,
	    fixing compiler warnings.

2018-08-13  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Drop commented midatrie stuff.

	* src/Makefile.am:
	  - Drop commented stuff on using ancient midatrie.

2018-08-10  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Normalize include guard macros in headers.

	* conv/conv.h:
	* conv/convfact.h:
	* conv/convkit.h:
	* conv/tis620.h:
	* conv/tischar.h:
	* conv/unichar.h:
	* conv/utf8.h:
	  - Normalize include guard macros based on file name.

2018-08-10  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Drop "#pragma once" in headers.

	It's not portable, and multiple include optimization
	in compilers just make "#ifndef" idiom have the same effect.

	* src/abswordseg.h:
	* src/filefilter.h:
	* src/filterhtml.h:
	* src/filterlambda.h:
	* src/filterlatex.h:
	* src/filterrtf.h:
	* src/filterx.h:
	* src/longwordseg.h:
	* src/maxwordseg.h:
	* src/wordstack.h:
	  - Drop "#pragma once".

2018-08-08  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Use HTTPS in linux.thai.net URLs

	* AUTHORS:
	  - Use HTTPS in thailatex and libthai project URLs.

2018-08-05  Theppitak Karoonboonyanan <theppitak@gmail.com>

	Update dictionary from thailatex GitHub.

	* data/Makefile.am:

src/dict.cpp
//
// dict.cpp - dictionary class
// Created: 2017-11-08
// Author: Theppitak Karoonboonyanan
//

#include "dict.h"

#include <stdio.h>
#include <string.h>
#include <sys/stat.h>

#ifdef WIN32
#define PATH_SEP "\\"
#else
#define PATH_SEP "/"
#endif
#if defined __WIN32__ || defined _WIN32 || defined _Windows
#if !defined S_ISDIR
#define S_ISDIR(m) (((m) & _S_IFDIR) == _S_IFDIR)
#define S_ISREG(m) (((m) & _S_IFREG) == _S_IFREG)
#endif
#endif

#define DICT_FILENAME "swathdic.tri"

//
// class Dict
//

bool
Dict::open (const char* dictPath)
{
  char*       trieBuff = 0;
  const char* triePath;
  struct stat sb;

  if (stat (dictPath, &sb) == -1)
    {
      perror (dictPath);
      return false;
    }

  if (S_ISDIR (sb.st_mode))
    {
      trieBuff = new char[strlen (dictPath) + 2 + sizeof (DICT_FILENAME)];
      sprintf (trieBuff, "%s" PATH_SEP DICT_FILENAME, dictPath);
      triePath = trieBuff;
    }
  else if (S_ISREG (sb.st_mode))
    {
      triePath = dictPath;
    }
  else
    {
      fprintf (stderr, "%s is not a directory or regular file\n", dictPath);
      return false;
    }

  dict = trie_new_from_file (triePath);

  if (trieBuff)
    {
      delete[] trieBuff;
    }

  return dict != 0;
}



src/wordseg.cpp
// ==========================================
// Author: Chaivit Poovanich
// October 14, 1998
// Modified by : Paisarn Charoenpornsawat
// 1999
// ==========================================
// New features.
//       1. Support many formats: LaTeX, HTML,
//               Text, RTF (for Opensource, Emacs 1999)
//       2. Remove garbage characters.
//               (for a. virach requirement 2000)
//       3. Select wordseg algorithms, Bi-gram
//           Maximal Matching, Longest Matching
//               (for Opensource, Emacs 1999)
// ===========================================


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <wchar.h>
#include <wctype.h>

#include "worddef.h"
#include "filefilter.h"
#include "filterx.h"
#include "longwordseg.h"
#include "maxwordseg.h"
#include "convutil.h"
#include "utils.h"

enum TextToken {
  TTOK_EOT    = -1,
  TTOK_WSPACE = 0,
  TTOK_THAI   = 1,
  TTOK_ETC    = 2,
};

static TextToken SplitToken (wchar_t** str, wchar_t* token);

static bool
OpenDict (Dict* dict, const char* dictpath)
{
  // Dict search order:
  // 1. dictpath, if not NULL, exit on failure
  // 2. ${SWATHDICT} env, if not NULL, continue on failure
  // 3. current dir, continue on failure
  // 4. WORDSEGDATADIR macro
  if (dictpath)
    {
      if (dict->open (dictpath))
        return true;
    }
  else
    {
      dictpath = getenv ("SWATHDICT");
      if (dictpath && dict->open (dictpath))
        return true;

      if (dict->open ("."))
        return true;

      //if (dict->open (WORDSEGDATA_DIR))
      //  return true;
    }

  return false;
}

static AbsWordSeg*
InitWordSegmentation (const char* method)
{
  if (method)
    {
      if (strcmp (method, "long") == 0)
        return new LongWordSeg;
      else if (strcmp (method, "max") == 0)
        return new MaxWordSeg;
    }

  // fallback
  return new MaxWordSeg;
}

static void
ExitWordSegmentation (AbsWordSeg* wseg)
{
  delete wseg;
}

static void
WordSegmentation (AbsWordSeg* wseg, const Dict* dict, const wchar_t* wbr,
                  const wchar_t* line, wchar_t* output, int outputSz)
{
  short int *seps = new short int [outputSz];
  int nSeps = wseg->WordSeg (dict, line, seps, outputSz);

  const wchar_t* pLine = line;
  int wbrLen = wcslen (wbr);
  for (int i = 0; i < nSeps; i++)
    {
      const wchar_t* wordEnd = line + seps[i];
      int outLen = wordEnd - pLine;
      if (outLen >= outputSz)
        break;
      wcsncpy (output, pLine, outLen);
      output[outLen] = 0;
      output += outLen;
      outputSz -= outLen;

      if (wbrLen >= outputSz)
        break;
      wcsncpy (output, wbr, wbrLen);
      output += wbrLen;
      outputSz -= wbrLen;

      pLine = wordEnd;
    }

  wcsncpy (output, pLine, outputSz);

  delete[] seps;
}

static void
Version ()
{
  //printf ("swath " VERSION "\n");
  printf
    ("Copyright (C) 2001-2006 Phaisarn Charoenpornsawat <phaisarn@nectec.or.th>\n"
     "Copyright (C) 2001-2014 Theppitak Karoonboonyanan <theppitak@gmail.com>\n"
     "License: GNU GPL version 2 or later <http://gnu.org/licenses/gpl-2.0.html>\n"
     "This is free software; you are free to change and redistribute it.\n"
     "There is NO WARRANTY, to the extent permitted by law.\n");
}

static void
Usage (int verbose)
{
  fprintf (stderr,
           "Usage: swath [mule] [-v|--verbose] [-b \"delimiter\"] [-d dict-dir]\n"
           "       [-f html|rtf|latex|lambda] [-m long|max] [-u {u|t},{u|t}] [-h|[-]-help]\n");
  if (verbose)
    {
      fprintf (stderr,
               "Options:\n"
               "\tmule : for use with mule\n"
               "\t-v   : verbose mode\n"
               "\t-b   : define a word delimiter string for the output\n"
               "\t-d   : specify dictionary path\n"
               "\t-f   : specify format of the input\n"
               "\t\thtml     : HTML file\n"
               "\t\trtf      : RTF file\n"
               "\t\tlatex    : LaTeX file\n"
               "\t\tlambda   : The input and output are same as latex, except that\n"
               "\t\t           the word delimiter is ^^^^200b\n"
               "\t-m   : choose word matching scheme when analyzing\n"
               "\t\tlong     : longest matching scheme\n"
               "\t\tmax      : maximal matching scheme\n"
               "\t-u   : specify encodings of input and output in 'i,o' form,\n"
               "\t       for input and output respectively, where 'i', 'o' is one of:\n"
               "\t\tu        : The input/output is in UTF-8\n"
               "\t\tt        : The input/output is in TIS-620\n"
               "\t-help: display this help message\n");
    }
}

#ifndef WORDSEGDATA_DIR
#define WORDSEGDATA_DIR "/usr/local/lib/wordseg"
#endif

int
main (int argc, char* argv[])
{
  const char* wbr = NULL;
  const char* dictpath = NULL;
  const char* method = NULL;
  const char* fileformat = NULL;
  const char* unicode = NULL;
  bool verbose = false;
  bool muleMode;
  bool thaiFlag;

  muleMode = false;
  for (int iargc = 1; iargc < argc; iargc++)
    {
      if (strcmp ("mule", argv[iargc]) == 0)
        {
          muleMode = true;
          verbose = false;
        }
      else if (strcmp ("-b", argv[iargc]) == 0 && iargc + 1 < argc)
        {
          wbr = argv[++iargc];
        }
      else if (strcmp ("-d", argv[iargc]) == 0 && iargc + 1 < argc)
        {
          dictpath = argv[++iargc];
        }
      else if (strcmp ("-f", argv[iargc]) == 0 && iargc + 1 < argc)
        {
          fileformat = argv[++iargc];
        }
      else if (strcmp ("-v", argv[iargc]) == 0 ||
               strcmp ("--verbose", argv[iargc]) == 0)
        {
          verbose = true;
        }
      else if (strcmp ("-m", argv[iargc]) == 0 && iargc + 1 < argc)
        {
          method = argv[++iargc];
        }
      else if (strcmp ("-u", argv[iargc]) == 0 && iargc + 1 < argc)
        {
          unicode = argv[++iargc];
        }
      else if (strcmp ("-V", argv[iargc]) == 0 ||
               strcmp ("--version", argv[iargc]) == 0)
        {
          Version ();
          return 0;
        }
      else if (strcmp ("-help", argv[iargc]) == 0 ||
               strcmp ("-h", argv[iargc]) == 0 ||
               strcmp ("--help", argv[iargc]) == 0)
        {
          Usage (1);
          return 1;
        }
      else
        {
          Usage (0);
          return 1;
        }
    }

  if (verbose)
    printf ("*** Word Segmentation ***\n");

  Dict dict;
  if (!OpenDict (&dict, dictpath))
    return 1;

  AbsWordSeg* wseg = InitWordSegmentation (method);
  if (!wseg)
    return 1;

  bool  isUniIn = false;
  bool  isUniOut = false;
  if (unicode != NULL)
    {                           //Option -u
      if (unicode[0] == 'u')
        {
          isUniIn = true;
        }
      if (unicode[2] == 'u')
        {
          isUniOut = true;
        }
    }

  wchar_t wLine[MAXLEN];
  wchar_t wsegOut[MAXLEN * 2];

  if (fileformat != NULL)
    {
      FilterX* FltX = FileFilter::CreateFilter (stdin, stdout,
                                                isUniIn, isUniOut,
                                                fileformat);
      if (FltX == NULL)
        {
          fprintf (stderr, "Invalid file format: %s\n", fileformat);
          ExitWordSegmentation (wseg);
          return 1;
        }
      while (FltX->GetNextToken (wLine, N_ELM (wLine), &thaiFlag))
        {
          if (!thaiFlag)
            {
              FltX->Print (wLine, thaiFlag);
              continue;
            }
          WordSegmentation (wseg, &dict, FltX->GetWordBreak(), wLine,
                            wsegOut, N_ELM (wsegOut));
          FltX->Print (wsegOut, thaiFlag);
        }
      delete FltX;
    }
  else
    {
      wchar_t* wcwbr_buff = NULL;
      const wchar_t* wcwbr = L"|";
      if (wbr)
        {
          wcwbr = wcwbr_buff = ConvStrDup (wbr, isUniOut);
        }
      while (!feof (stdin))
        {
          if (verbose)
            printf ("Input : ");

          if (!ConvGetS (wLine, N_ELM (wLine), stdin, isUniIn))
            break;
          int len = wcslen (wLine);
          if (L'\n' == wLine [len - 1])
            wLine [--len] = 0;

          if (verbose)
            printf ("Output: ");

          if (0 == wLine[0])
            {
              printf ("\n");
              continue;
            }

          int tokenFlag;
          wchar_t* startStr = wLine;
          wchar_t wToken[MAXLEN];

          while ((tokenFlag = SplitToken (&startStr, wToken)) != TTOK_EOT)
            {
              switch (tokenFlag)
                {
                  case TTOK_THAI:
                    WordSegmentation (wseg, &dict, wcwbr, wToken, wsegOut,
                                      N_ELM (wsegOut));
                    ConvPrint (stdout, wsegOut, isUniOut);
                    break;

                  case TTOK_WSPACE:
                  case TTOK_ETC:
                  default:
                    ConvPrint (stdout, wToken, isUniOut);
                    break;
                }
              if (muleMode)
                {
                  ConvPrint (stdout, wcwbr, isUniOut);
                }
            }
          printf ("\n");
        }
      if (wcwbr_buff)
        free (wcwbr_buff);
    }

  ExitWordSegmentation (wseg);

  return 0;
}

static TextToken
SplitToken (wchar_t** str, wchar_t* token)
{
  if (**str == 0)
    return TTOK_EOT;

  if (iswspace (**str))
    {
      while (**str != 0 && iswspace (**str))
        {
          *token++ = *(*str)++;
        }
      *token = 0;
      return TTOK_WSPACE;
    }
  else if (isThaiUni (**str))
    {
      while (**str != 0 && isThaiUni (**str))
        {
          *token++ = *(*str)++;
        }
      *token = 0;
      return TTOK_THAI;
    }
  else
    {
      while (**str != 0 && !iswspace (**str) && !isThaiUni (**str))
        {
          *token++ = *(*str)++;
        }
      *token = 0;
      return TTOK_ETC;
    }
}

